File: SwimSharp/SwimSharp/SCALA/DONE/Messages.scala
package fi.jihartik.swim

import akka.util.ByteString
import spray.json._


trait UdpMessage {
  def toMessageString: String
  def toByteString: ByteString = ByteString(toMessageString)
}
trait FailureDetectionMessage extends UdpMessage
trait ClusterStateMessage

case class Ping(seqNo: Long) extends FailureDetectionMessage {
  def toMessageString = s"0$seqNo"
}
case class IndirectPing(seqNo: Long, target: Member) extends FailureDetectionMessage {
  import JsonSerialization._
  def toMessageString = s"1$seqNo ${target.toJson.compactPrint}"
}
case class Ack(seqNo: Long) extends FailureDetectionMessage {
  def toMessageString = s"2$seqNo"
}
abstract class MemberStateMessage(msgType: Int) extends UdpMessage with ClusterStateMessage {
  import JsonSerialization._
  def member: Member
  def toMessageString = s"$msgType${member.toJson.compactPrint}"
}
case class AliveMember(member: Member) extends MemberStateMessage(3)
case class SuspectMember(member: Member) extends MemberStateMessage(4)
case class DeadMember(member: Member) extends MemberStateMessage(5)
case class CompoundUdpMessage(messages: List[UdpMessage]) extends UdpMessage with ClusterStateMessage {
  def toMessageString = "6" ++ messages.map { msg =>
    val msgBody = msg.toMessageString
    val msgHeader = msgBody.size + " "
    msgHeader ++ msgBody
  }.foldLeft("")(_ + _)
}
case class NewMembers(members: List[Member]) extends ClusterStateMessage


object UdpMessage {
  import JsonSerialization._

  def apply(payload: ByteString): UdpMessage = apply(payload.decodeString("UTF-8"))

  def apply(decoded: String): UdpMessage = {
    val msgType = decoded.take(1).toInt
    val message = decoded.drop(1)
    msgType match {
      case 0 => Ping(message.toLong)
      case 1 => IndirectPing(message.takeWhile(_ != ' ').toLong, message.dropWhile(_ != ' ').drop(1).asJson.convertTo[Member])
      case 2 => Ack(message.toLong)
      case 3 => AliveMember(message.asJson.convertTo[Member])
      case 4 => SuspectMember(message.asJson.convertTo[Member])
      case 5 => DeadMember(message.asJson.convertTo[Member])
      case 6 => parseCompoundMessage(message)
    }
  }

  private def parseCompoundMessage(message: String) = {
    def parseNextMessage(message: String, messages: List[UdpMessage]): List[UdpMessage] = {
      if(message != "") {
        val (size, rest) = message.splitAt(message.indexOf(" "))
        val messageBody = rest.drop(1).take(size.toInt)
        parseNextMessage(rest.drop(1).drop(size.toInt), apply(messageBody) :: messages)
      } else {
        messages
      }
    }
    CompoundUdpMessage(parseNextMessage(message, Nil))
  }
}

File: SwimSharp/SwimSharp/Messages.cs
using Akka.IO;
using Newtonsoft.Json;
using SwimSharp.Extensions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SwimSharp
{
    internal abstract class UdpMessage
    {
        internal abstract String ToMessageString();
        internal ByteString ToByteString() => ByteString.FromString(ToMessageString());

        internal static UdpMessage Apply(ByteString payload)
        {
            return Apply(payload.ToString(Encoding.UTF8));
        }

        internal static UdpMessage Apply(String decoded)
        {
            var msgType = int.Parse(new String(decoded.Take(1).ToArray()));
            var message = new String(decoded.Skip(1).ToArray());
            switch(msgType)
            {
                case 0:
                    return new Ping(long.Parse(message));
                case 1:
                    var seqNo = long.Parse(new String(message.TakeWhile(x => x != ' ').ToArray()));
                    var idx = message.IndexOf(' ') + 1;
                    var member = JsonConvert.DeserializeObject<Member>(message.Substring(idx));
                    return new IndirectPing(seqNo, member);
                case 2:
                    return new Ack(long.Parse(message));
                case 3:
                    var member = JsonConvert.DeserializeObject<Member>(message);
                    return new AliveMember(member);
                case 4:
                    var member = JsonConvert.DeserializeObject<Member>(message);
                    return new SuspectMember(member);
                case 5:
                    var member = JsonConvert.DeserializeObject<Member>(message);
                    return new DeadMember(member);
                case 6:
                    return ParseCompoundMessage(message);
                default:
                    throw new InvalidProgramException("Out of range value for msgType in UdpMessage.Apply()")
            }
        }

        private CompoundUdpMessage ParseCompoundMessage(string messageOuter)
        {
            List<UdpMessage> ParseNextMessage(string message, List<UdpMessage> messages)
            {
                if(message != "")
                {
                    var (size, rest) = message.SplitAt(message.IndexOf(" "));
                    var messageBody = new String(rest.Skip(1).Take(int.Parse(size)).ToArray());
                    messages.Add(Apply(messageBody));
                    return ParseNextMessage(
                        new String(rest.Skip(1).Skip(int.Parse(size)).ToArray()), messages);
                }
                else
                {
                    return messages;
                }
            }
            return new CompoundUdpMessage(ParseNextMessage(messageOuter, new List<UdpMessage>()));
        }
    }

    internal abstract class FailureDetectionMessage : UdpMessage
    {
    }

    internal interface ClusterStateMessage
    {
    }

    internal class Ping : FailureDetectionMessage
    {
        internal Ping(long seqNo)
        {
            this.SeqNo = seqNo;
        }

        internal long SeqNo { get; private set; }

        internal override string ToMessageString()
        {
            return $"0{SeqNo}";
        }
    }

    internal class IndirectPing : FailureDetectionMessage
    {
        internal IndirectPing(long seqNo, Member target)
        {
            this.SeqNo = seqNo;
            this.Target = target;
        }

        internal long SeqNo { get; private set; }
        internal Member Target { get; private set; }

        internal override string ToMessageString()
        {
            var targetJsonString = JsonConvert.SerializeObject(Target, Formatting.None);
            return $"1{SeqNo} {targetJsonString}";
        }
    }

    internal class Ack : FailureDetectionMessage
    {
        internal Ack(long seqNo)
        {
            this.SeqNo = seqNo;
        }

        internal long SeqNo { get; private set; }

        internal override string ToMessageString()
        {
            return $"2{SeqNo}";
        }
    }

    internal abstract class MemberStateMessage : UdpMessage, ClusterStateMessage
    {
        public MemberStateMessage(Member member, int msgType)
        {
            this.Member = member;
            this.MsgType = msgType;
        }

        internal Member Member { get; private set; }
        internal int MsgType { get; private set; }

        internal override string ToMessageString()
        {
            var targetJsonString = JsonConvert.SerializeObject(Member, Formatting.None);
            return $"{MsgType}{targetJsonString}";
        }
    }

    internal class AliveMember : MemberStateMessage
    {
        public AliveMember(Member member) : base(member, 3) { }
    }

    internal class SuspectMember : MemberStateMessage
    {
        public SuspectMember(Member member) : base(member, 4) { }
    }


    internal class DeadMember : MemberStateMessage
    {
        public DeadMember(Member member) : base(member, 5) { }
    }

    internal class CompoundUdpMessage : UdpMessage, ClusterStateMessage
    {
        public CompoundUdpMessage(List<UdpMessage> messages)
        {
            this.Messages = messages;
        }

        public List<UdpMessage> Messages { get; private set; }

        internal override string ToMessageString()
        {
            var messages = Messages.Select(msg =>
                           {
                               var msgBody = msg.ToMessageString();
                               var msgHeader = msgBody.Length + " ";
                               return msgHeader + msgBody;

                           });
            var joined = String.Join("", messages);
            return $"6{joined}";
        }
    }

    internal class NewMembers : ClusterStateMessage
    {
        public NewMembers(List<Member> members)
        {
            this.Members = members;
        }

        public List<Member> Members { get; private set; }
    }
}



